syft/pkg/cataloger/golang/licenses.go

```diff
@@ -68,38 +67,64 @@ func modCacheResolver(modCacheDir string) source.FileResolver {
 			log.Tracef("unable to open mod cache directory: %s, skipping mod cache resolver", modCacheDir)
 			r = source.NewMockResolverForPaths()
 		} else {
-			r = source.NewDeferredResolverFromSource(func() (source.Source, error) {
-				return source.NewFromDirectory(modCacheDir)
-			})
+			r = source.NewUnindexedDirectoryResolver(modCacheDir)
 		}
 	}
 
-	modCacheResolvers[modCacheDir] = r
-
 	return r
 }
 
 func (c *goLicenses) getLicenses(resolver source.FileResolver, moduleName, moduleVersion string) (licenses []string, err error) {
-	moduleName = processCaps(moduleName)
-
 	licenses, err = findLicenses(resolver,
-		fmt.Sprintf(`**/go/pkg/mod/%s@%s/*`, moduleName, moduleVersion),
+		fmt.Sprintf(`**/go/pkg/mod/%s@%s/*`, processCaps(moduleName), moduleVersion),
 	)
 
-	if c.searchLocalModCacheLicenses && err == nil && len(licenses) == 0 {
+	search := c.opts.searchLocalModCacheLicenses || c.opts.searchRemoteLicenses
+	dir := fmt.Sprintf("%s@%s", processCaps(moduleName), moduleVersion)
+	glob := fmt.Sprintf("%s/*", dir)
+
+	if search && err == nil && len(licenses) == 0 {
```

Comment:
I think there are opportunities to make these code blocks a little more grokkable by separating some of the compound conditional, using more guard clauses, and decomposing functions into smaller pieces. For example:
```golang

func (c *goLicenses) getLicenses(resolver source.FileResolver, moduleName, moduleVersion string) (licenses []string, err error) {
	// look in the target resolver...
	licenses, err = findLicenses(resolver,
		fmt.Sprintf(`**/go/pkg/mod/%s@%s/*`, processCaps(moduleName), moduleVersion),
	)
	if err != nil || len(licenses) > 0 {
		return requireCollection(licenses), err
	}

	// look in the local host mod cache...
	licenses, err = c.getLicensesFromLocal(moduleName, moduleVersion)
	if err != nil || len(licenses) > 0 {
		return requireCollection(licenses), err
	}

	// fetch the module remotely and search...
	licenses, err = c.getLicensesFromRemote(moduleName, moduleVersion)
	if err != nil || len(licenses) > 0 {
		return requireCollection(licenses), err
	}

	return requireCollection(licenses), err
}

func (c *goLicenses) getLicensesFromRemote(moduleName, moduleVersion string) ([]string, error) {
	if !c.opts.searchRemoteLicenses {
		return nil, nil
	}

	if err := c.fetchModuleFromRemote(moduleName, moduleVersion); err != nil {
		log.Tracef("remote proxy walk failed for: %s@%s", moduleName, moduleVersion)
		return nil, err
	}

	return findLicenses(c.localModCacheResolver, moduleSearchGlob(moduleName, moduleVersion))

}

func (c *goLicenses) fetchModuleFromRemote(moduleName, moduleVersion string) error {
	proxies := remotesForModule(c.opts.proxies, c.opts.noProxy, moduleName)

	var fsys fs.FS
	fsys, err := getModule(c.progress, proxies, moduleName, moduleVersion)
	if err != nil {
		return err
	}

	// populate the mod cache with the results
	return fs.WalkDir(fsys, ".", func(filePath string, d fs.DirEntry, _ error) error {
		if d.IsDir() {
			return nil
		}
		f, err := fsys.Open(filePath)
		if err != nil {
			return err
		}
		return c.localModCacheResolver.Write(
			source.NewLocation(
				path.Join(
					moduleDir(moduleName, moduleVersion),
					filePath),
			),
			f)
	})

}

func (c *goLicenses) getLicensesFromLocal(moduleName, moduleVersion string) ([]string, error) {
	isEnabled := c.opts.searchLocalModCacheLicenses || c.opts.searchRemoteLicenses

	if !isEnabled {
		return nil, nil
	}

	// if we're running against a directory on the filesystem, it may not include the
	// user's homedir / GOPATH, so we defer to using the localModCacheResolver
	return findLicenses(c.localModCacheResolver, moduleSearchGlob(moduleName, moduleVersion))
}

func moduleDir(moduleName, moduleVersion string) string {
	return fmt.Sprintf("%s@%s", processCaps(moduleName), moduleVersion)
}

func moduleSearchGlob(moduleName, moduleVersion string) string {
	return fmt.Sprintf("%s/*", moduleDir(moduleName, moduleVersion))
}

func requireCollection(licenses []string) []string {
	if licenses == nil {
		return []string{}
	}
	return licenses
}

```

Consider this comment as optional (no action required), but it is encouraged.